### PR TITLE
Updated scoring model schema

### PR DESCRIPTION
**Pull request by lmacielvieira**
### Description

Applied pertinent updates to support new ML models and updated scoring model schema to support behavior similar to molecular databases #1491 #1509 

*** analysis_version field kept for backwards compatibility

#### How to test it
1 - Run migration
2 - Run metaspace/docker/sm-engine/install-scoring-model.sh script to install ML models
3 - Change old ML model (v3_default) column is_archived to 'true'

image
image
image

### CODE REVIEW

File: metaspace/engine/sm/engine/annotation_lithops/run_fdr.py [hunk 1]
```diff
@@ -42,7 +42,12 @@ def _run_fdr_for_db(db_data_cobject: CObj[DbFDRData], *, storage: Storage):
         return db_data['id'], results_df
 
     logger.info('Estimating FDRs...')
-    scoring_model = load_scoring_model(ds_config['fdr'].get('scoring_model'))
+    logger.info(ds_config['fdr'].get('scoring_model'))
+    logger.info(ds_config['fdr'].get('scoring_model_id'))
+    logger.info(ds_config['fdr'].get('scoring_model_version'))
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/engine/annotation_lithops/run_fdr.py:47]: please change it for one line, like
```
logger.info(f'Scoring model: MSM')
or
logger.info(f'Scoring model: {__name__} / {__version__}')
```

File: metaspace/engine/scripts/import_scoring_model.py [hunk 2]
```diff
@@ -12,29 +12,67 @@
...
+        parser.add_argument(
+            'bucket',
+            type=str,
+            nargs='?',
+            default=None,
+            help='S3 or MinIO bucket to upload to. Optional for original models.',
+        )
+    else:
+        parser.add_argument('model', type=str, help='Path to a CBM model file.')
+        parser.add_argument('bucket', type=str, help='S3 or MinIO bucket to upload to')
```
Comment (sergii-mamedov) [inline metaspace/engine/scripts/import_scoring_model.py:75]: I think you over complicated the logic a little.
I would not create a `pre_parser` and make the `model_type` argument mandatory. The parameters `model` and `bucket` setting up by default to `None`, and only if `model_type == 'catboost'` would I check for these arguments.

File: metaspace/engine/sm/engine/annotation/scoring_model.py [hunk 3]
```diff
@@ -88,9 +106,61 @@ def score(
...
+    data = DB().select_one_with_fields(
+        'SELECT id, name, version, type FROM scoring_model WHERE name = %s AND version = %s',
+        params=(name, version),
+    )
+    if not data:
+        raise SMError(f'ScoringModel not found: {name}')
+    return ScoringModel(**data)
+
+
+def find_by_id(id_: int) -> ScoringModel:
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/engine/annotation/scoring_model.py:145]: Do we still need all these functions? `find_original`, `find_by_name_version`, `find_by_id`

File: metaspace/engine/sm/engine/dataset.py [hunk 4]
```diff
@@ -242,14 +247,23 @@ def generate_ds_config(
...
+        analysis_version = analysis_version or 1
+    else:
+        analysis_version = 1 if model_type == 'original' else 3
     iso_params = _get_isotope_generation_from_metadata(metadata)
     default_adducts, charge, isocalc_sigma, instrument = iso_params
-    default_scoring_model = 'v3_default' if analysis_version >= 3 else None
+
+    scoring_model = None  # original MSM backwards compatibility
+    if scoring_model_id and model_type != 'original':
+        scoring_model = find_by_id(scoring_model_id)
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/engine/dataset.py:266]: I think it is not needed anymore. Or not?

File: metaspace/engine/sm/fdr_engineering/run_train_eval.py [hunk 5]
```diff
@@ -231,6 +231,7 @@ def eval_model(model, metrics_df):
 #%% Save model to S3
 
 MODEL_NAME = 'v3_default'
```
Comment (sergii-mamedov) [inline metaspace/engine/sm/fdr_engineering/run_train_eval.py:233]: It it still relevant?

**Review by sergii-mamedov — CHANGES_REQUESTED** [review]

File: metaspace/python-client/metaspace/sm_annotation_utils.py [hunk 6]
```diff
@@ -693,6 +711,36 @@ def map_database_to_id(self, database: Union[int, str, Tuple[str, str]]):
...
+        if isinstance(scoring_model, str) and re.match(r'^\d+$', scoring_model):
+            return int(scoring_model)
+
+        scoring_model_docs = self.get_visible_scoring_models()
+        scoring_model_name_id_map = defaultdict(list)
+        for sm in scoring_model_docs:
+            scoring_model_name_id_map[(sm['name'], sm['version'])].append(sm['id'])
+
+        if isinstance(scoring_model, tuple):
+            db_name, db_version = scoring_model
```
Comment (sergii-mamedov) [inline metaspace/python-client/metaspace/sm_annotation_utils.py:727]: please change `db_name`, ... to `model_name` or something like this

**Review by sergii-mamedov — CHANGES_REQUESTED** [review]

File: metaspace/engine/sm/engine/dataset.py [hunk 7]
```diff
@@ -242,14 +247,23 @@ def generate_ds_config(
...
     compute_unused_metrics=None,
-    scoring_model=None,
+    scoring_model_id=None,
+    model_type=None,
 ) -> DSConfig:
     # The kwarg names should match FLAT_DS_CONFIG_KEYS
 
-    analysis_version = analysis_version or 1
+    # 1 - original, 2 - v2 (discontinued), 3 - ML (catboost or other != original)
+    # analysis_version kept for compatibility with old datasets and client
```
[the remark this answers is not in the document]
Comment (lmacielvieira) [reply metaspace/engine/sm/engine/dataset.py]: I've removed as parameter, but kept the backward compatibility by setting it according to model typ.

File: metaspace/engine/sm/engine/annotation_lithops/run_fdr.py (hunk 1 above)
Comment (lmacielvieira) [reply metaspace/engine/sm/engine/annotation_lithops/run_fdr.py]: I've removed it.

File: metaspace/engine/sm/engine/dataset.py (hunk 4 above)
Comment (lmacielvieira) [reply metaspace/engine/sm/engine/dataset.py]: Removed find_original, but find_by_id is necessary to get model_type as analysis_version parameter was discontinued

File: metaspace/engine/sm/fdr_engineering/run_train_eval.py (hunk 5 above)
Comment (lmacielvieira) [reply metaspace/engine/sm/fdr_engineering/run_train_eval.py:233]: Should we ask Bishoy if he still uses to evaluate the model?

**Review by sergii-mamedov — APPROVED** [review]
LGTM

**Pull request merged**